coderd/workspaces.go

```diff
@@ -294,28 +294,39 @@ func (api *API) postWorkspacesByOrganization(rw http.ResponseWriter, r *http.Req
 		return
 	}
 
-	workspace, err := api.Database.GetWorkspaceByOwnerIDAndName(ctx, database.GetWorkspaceByOwnerIDAndNameParams{
+	workspaces, err := api.Database.GetWorkspaces(ctx, database.GetWorkspacesParams{
 		OwnerID: user.ID,
-		Name:    createWorkspace.Name,
 	})
-	if err == nil {
-		// If the workspace already exists, don't allow creation.
-		httpapi.Write(ctx, rw, http.StatusConflict, codersdk.Response{
-			Message: fmt.Sprintf("Workspace %q already exists.", createWorkspace.Name),
-			Validations: []codersdk.ValidationError{{
-				Field:  "name",
-				Detail: "This value is already in use and should be unique.",
-			}},
-		})
-		return
-	}
 	if err != nil && !errors.Is(err, sql.ErrNoRows) {
 		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
 			Message: fmt.Sprintf("Internal error fetching workspace by name %q.", createWorkspace.Name),
 			Detail:  err.Error(),
 		})
 		return
 	}
+	for _, workspace := range workspaces {
+		if workspace.Name == createWorkspace.Name {
+			// If the workspace already exists, don't allow creation.
+			httpapi.Write(ctx, rw, http.StatusConflict, codersdk.Response{
+				Message: fmt.Sprintf("Workspace %q already exists.", createWorkspace.Name),
+				Validations: []codersdk.ValidationError{{
+					Field:  "name",
+					Detail: "This value is already in use and should be unique.",
+				}},
+			})
+			return
+		}
+	}
+
+	// make sure the user has not hit their quota limit
+	e := *api.WorkspaceQuotaEnforcer.Load()
+	canCreate := e.CanCreateWorkspace(len(workspaces))
+	if !canCreate {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: fmt.Sprintf("User workspace limit of %d is already reached.", e.UserWorkspaceLimit()),
```

Comment:
Could `WorkspaceQuotaEnforcer` be converted to an atomic integer instead? This seems to be only be used against the count.

eg.
```
WorkspaceLimit atomic.Pointer[int64]
```
Then if the limit is zero, we just don't do this behavior.

Reply:
I wanted to keep the `CanCreateWorkspace` check behind the enforcer to keep the logic from leaking into the workspace route so we can change it easier when we want to add more features to this down the road. But yes technically we can just make it an int and do the logic in this route.

Reply:
Nah, that makes sense to me. Maybe worth a comment, but it's fine even if not.